Add StarRating (#12)
30-seconds/30-seconds-of-react: snippets/StarRating.md

Create two methods, `hoverOver` and `setRating`, that take an event as argument and update `state.selected` and `state.rating` according to it, bind them both to the component's context.
In the `render()` method, create a `<div>` to wrap the `<Star>` components, which are created using `Array.prototype.map` on an array of 5 elements, created using `Array.from`, and handle the `onMouseLeave` event to set `state.selection` to `0`, the `onClick` event to set
the `state.rating` and the `onMouseOver` event to set `state.selection` to the `star-id` attribute of the `event.target` respectively. 
Finally, pass the appropriate values to each `<Star>` component (`starId` and `marked`).

```jsx
const StarRating = (function() {
  function Star({ marked, starId }) {
    return (
      <span star-id={starId} style={{ color: '#ff9933' }} role='button'>
        {marked ? '\u2605' : '\u2606'}
      </span>
    );
  }

  return class StarRating extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        rating: typeof props.rating == 'number' ? props.rating : 0,
        selection: 0
      };
      this.hoverOver = this.hoverOver.bind(this);

>       this.handleClick = this.handleClick.bind(this);

Reviewer: We can use the binding to also do the hover out

```suggestion
      this.hoverOut = this.hoverOver.bind(this, null);
      this.handleClick = this.handleClick.bind(this);
```

    }
    hoverOver(event) {
      let val = 0;
      if (event && event.target && event.target.getAttribute('star-id'))
        val = event.target.getAttribute('star-id');
      this.setState(state => ({ selection: val }));
    }
    handleClick(event) {
      const val = event.target.getAttribute('star-id') || this.state.rating;
      this.setState(state => ({ rating: val }));
    }
    render() {
      return (
        <div

>           onMouseOut={() => this.hoverOver(null)}

Reviewer: If we change the way we're doing the hoverOut we can just use it here

```suggestion
          onMouseOut={this.hoverOut}
```

          onClick={this.handleClick}
          onMouseOver={this.hoverOver}
        >

>           {Array.from({ length: 5 }, (v, i) => i + 1).map(v => (

Reviewer: We can move the length to a `prop` so that it can be dynamically set at use, but if we just want a static five then this one works just fine.

Author: I think the snippet is already complex enough, adding a prop for that would only make it harder. Let's leave that for the advanced version.

Reviewer: then we'll leave it at a basic five

>           {Array.from({ length: 5 }, (v, i) => i + 1).map(v => (

Reviewer: We can avoid an additional map call by just using `i` directly. We could also increment `i`, but `i+1` is also short enough to just retype and have the compiler de-roll it(will do so for small repeated values)

```suggestion
          {Array.from({ length: 5 }, (v, i) => (
          <Star
              starId={i+1}
              key={`star_${i+1} `}
              marked={
                this.state.selection
                  ? this.state.selection >= i+1
                  : this.state.rating >= i+1
              }
            />
```

>           {Array.from({ length: 5 }, (v, i) => i + 1).map(v => (

Reviewer: We can avoid an additional map call by just using `i` directly. We could also increment `i`, but `i+1` is also short enough to just retype and have the compiler de-roll it(will do so for small repeated values)

```suggestion
          {Array.from({ length: 5 }, (v, i) => (
```

            <Star

>               starId={v}

Reviewer: ```suggestion
              starId={i+1}
```

>               key={`star_${v} `}

Reviewer: ```suggestion
              key={`star_${i+1} `}
```

              marked={
                this.state.selection

>                   ? this.state.selection >= v

Reviewer: ```suggestion
                  ? this.state.selection >= i+1
```

>                   : this.state.rating >= v

Reviewer: ```suggestion
                  : this.state.rating >= i+1
```
